File: pdf.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import print_function

import json
import os
import sys

import fitz


def bookmarks_for_pdf(file_name):
    """
    Get bookmarks for PDF using PyMuPDF

    :param file_name: path to PDF file
    :return: parsed bookmarks
    """
    if not os.path.isfile(file_name) and not file_name.endswith('.pdf'):
        print("Expected PDF file.")
        sys.exit()

    return get_bookmarks(file_name)


def bookmarks_to_json(bookmarks):
    """
    Converts bookmarks from list of tuples to dictionary, and saves to JSON file

    :param bookmarks: bookmarks
    :return:
    """
    bookmark_list = []

    for bookmark in bookmarks:
        bookmark_dict = {
            "title": bookmark[0],
            "section": bookmark[1],
            "destination": bookmark[2]
        }
        bookmark_list.append(bookmark_dict)

    print(json.dumps(bookmark_list, indent=4))


def get_cover(pdf_file):
    """

    :param pdf_file:
    :return:
    """
    print("TODO %s" % pdf_file)


def get_bookmarks(pdf_file):
    """
    Get bookmarks from file and parse with parse_bookmarks()
    :param pdf_file: path to file
    :return: parsed bookmarks
    """
    doc = fitz.open(pdf_file)
    toc = doc.getToC(doc)

    bookmarks = parse_bookmarks(toc)

    if bookmarks:
        return bookmarks
    else:
        return None


def parse_bookmarks(raw_bookmarks):
    """
    Parse bookmarks from PyMuPDF, tuple of (title, section, destination)

    :param raw_bookmarks: list of lists from PyMuPDF
    :return: parsed bookmarks
    """
    sections = {}
    parsed_bookmarks = []

    for bookmark in raw_bookmarks:
        try:
            level = int(bookmark[0])
            title = bookmark[1].strip()
            destination = str(bookmark[2]).strip()

            section = ''
            sections[level] = title

            if level == 1:
                section = None
            else:
                current_level = level
                while current_level > 1:
                    if section == '':
                        section = sections[current_level - 1]
                    else:
                        section = (sections[current_level - 1] +
                                   ' > ' + section)

                    current_level -= 1

            if section is not None:
                section = section

            new_bookmark = (title, section, destination)
            parsed_bookmarks.append(new_bookmark)
        except Exception as e:
            # some bookmarks w/negative number or possibly others?
            print('error %s' % e.message)
            continue

    return parsed_bookmarks


def main():
    """

    :return:
    """
    args = sys.argv

    if not len(args) == 3:
        print("Expected three arguments: [command] [PDF file]")
    else:
        if args[1] == 'bookmarks':
            bookmarks = bookmarks_for_pdf(args[2])
            bookmarks_to_json(bookmarks)


if __name__ == '__main__':
    main()

File: pdf.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os/exec"
	"strings"

	"github.com/gen2brain/go-fitz"
)

type FileBookmark struct {
	Title       string `json:"title"`
	Section     string `json:"section"`
	Destination string `json:"destination"`
	Uri         string
}

// FileBookmarks: for EPUB and PDF file, using go-fitz
func FileBookmarks(file string) ([]*FileBookmark, error) {
	doc, err := fitz.New(file)
	if err != nil {
		fmt.Println("error:", err)
	}
	defer doc.Close()

	outlines, err := doc.ToC()
	if err != nil {
		fmt.Println("error:", err)
	}
	return parseBookmarks(outlines), err
}

// FileBookmarks: for EPUB and PDF file, using go-fitz
func FileMetadata(file string) map[string]string {
	doc, err := fitz.New(file)
	if err != nil {
		fmt.Println("error:", err)
	}
	defer doc.Close()
	return doc.Metadata()
}

// FBookmarks for EPUB and PDF file, using Python script ./pdf.py
func bookmarksForPDF(file string) ([]*FileBookmark, error) {
	cmdArgs := []string{"FileBookmarks", file}

	output, err := exec.Command("./pdf.py", cmdArgs...).Output()
	if err != nil {
		log.Fatal(err)
	}

	// JSON stdout as []bytes, convert before return
	var bookmarks []*FileBookmark
	bookmarks, err = bookmarksFromJson(output)

	return bookmarks, err
}

// Take JSON []bytes and return as slice of FileBookmark structs
func bookmarksFromJson(jsonBytes []byte) ([]*FileBookmark, error) {
	var bookmarks []*FileBookmark

	err := json.Unmarshal(jsonBytes, bookmarks)
	if err != nil {
		fmt.Println(err)
	}
	return bookmarks, err
}

// Parse bookmarks from go-fitz
func parseBookmarks(outline []fitz.Outline) []*FileBookmark {
	sections := []string{"", "", "", "", "", "", "", "", "", "", "", "", ""}
	var parsedBookmarks []*FileBookmark
	var page int
	var currentLevel int
	var section string
	var title string

	for _, bookmark := range outline {
		if bookmark.Page != -1 {
			page = bookmark.Page + 1
		} else {
			page = -1
		}
		title = strings.TrimSpace(bookmark.Title)

		section = ""
		sections[bookmark.Level] = title

		if bookmark.Level == 1 {
			section = ""
		} else {
			currentLevel = bookmark.Level
			for currentLevel > 1 {
				if section == "" {
					section = sections[currentLevel-1]
				} else {
					section = fmt.Sprintf("%s > %s", sections[currentLevel-1], section)
				}
				currentLevel -= 1
			}
		}
		newBookmark := FileBookmark{
			Title:       title,
			Section:     section,
			Destination: fmt.Sprintf("%d", page),
			Uri:         strings.TrimSpace(bookmark.URI),
		}
		parsedBookmarks = append(parsedBookmarks, &newBookmark)
	}
	return parsedBookmarks
}
